### DISCUSSION TITLE
DBCP 종류 및 특성 비교

### DISCUSSION THREAD

**Discussion by joowankim** [start-with-kotlin/kotlin#14] [General]
DBCP 종류 및 특성 비교

**Comment by parksjin01** [comment]
**[HikariCP](https://github.com/brettwooldridge/HikariCP)**

**Reply by parksjin01** [reply]
### HikariCP의 주요 특징
- Fast
- Reliable

### 기존 DBCP와의 차이점
- 기존 DBCP는 주기적으로 풀안의 커넥션으로 `select 1` 등의 테스트 쿼리를 날려서 커넥션이 유효한지 확인했다.
- 주기적으로 커넥션을 날리면 DB 서버에 의해서 커넥션이 끊어지는 것도 막을 수 있었다.
- HikariCP는 커넥션 풀에 있는 커넥션으로 테스트 쿼리를 날리지 않는다.
- HikariCP는 client가 `getConnection()` 메소드 호출시 JDBC4의 `isValid()` 메소드로 커넥션의 유효성을 검증한다음 client에게 커넥션을 반환한다.
- 만약, 사용하는 JDBC가 JDBC4를 지원하지 않는다면 `connectionTestQuery` 설정값으로 지정한 쿼리를 사용한다.
- HikariCP는 주기적으로 테스트 쿼리를 날려서 DB서버에 의해 커넥션이 끊어지는 것을 막는 대신 `maxLifeTime` 값보다 오랫동안 풀에 있었던 커넥션은 아직 유효하더라도 커넥션 풀에서 제거하고 새로운 커넥션을 추가한다.
- `maxLifeTime`이 DB서버에 설정된 커넥션 타임아웃 시간보다 작게 설정되면 DB서버에 의해서 커넥션이 끊어지는 문제를 해결할 수 있다.\
- HikariCP는 DBCP단에서 PreparedStatement 캐싱을 지원하지 않는다.
  - JDBC가 이미 PreparedStatement 캐싱을 지원하고, DBCP단에서 캐싱을 지원하는 경우 커넥션별로 캐싱을 해야하는데, JDBC가 제공하는 캐싱은 커넥션간 공유가 가능하기 때문
- Statement 로깅도 지원하지 않는다.
  - 로깅도 캐싱과 동일하게 대부분의 JDBC가 지원하기 때문에 JDBC에 위임한다.

### HikariCP의 장점
1. 빠르다.
  - HikariCP Github 의 README에 성능 벤치마크가 나와있다.
2. 많은 프로젝트에서 사용하고 있다. 
  - Java에서는 대부분 프로젝트가 Spring, Spring Boot를 사용하는데 Spring Boot 2 이상은 기본 DBCP로 HikariCP를 사용한다.

### 함께 읽어보면 좋을 문서들
- https://github.com/brettwooldridge/HikariCP
- https://brunch.co.kr/@jehovah/24
- https://brunch.co.kr/@jehovah/25
- https://woowabros.github.io/experience/2020/02/06/hikaricp-avoid-dead-lock.html
- https://woowabros.github.io/experience/2020/02/06/hikaricp-avoid-dead-lock-2.html
- https://kakaocommerce.tistory.com/45

**Comment by joowankim** [comment]
[DBCP 간략히 정리](https://github.com/start-with-kotlin/kotlin/blob/master/DBCP/DBCP.md)
[Apache Commons 커넥션 풀 내의 커넥션 개수 설정 정리](https://github.com/start-with-kotlin/kotlin/blob/master/DBCP/Apache%20Commons%20DBCP/Apache%20Commons%20DBCP.md)

**Reply by joowankim** [reply]
[여러 DBCP 비교글](https://medium.com/@jeevanpaatil/database-connection-pool-analysis-92d50ba4bd06)

**Comment by joowankim** [comment]
hikariCP
Q. 커넥션이 소멸되기 바로 전 시점에 요청을 처리하기 위해 저 케넥션 선택이 됐을 때
- 사용중인 커넥션은 소멸 시키지 않는다.
- 사용이 끝나면 바로 그 커넥션은 소멸된다.

**Comment by parksjin01** [comment]
Hibernate
